lib/TH/THGeneral.c

```diff
@@ -139,10 +141,14 @@ static long getAllocSize(void *ptr) {
  *     soft max by 40%
  */
 static void maybeTriggerGC(long curHeapSize) {
-  if(torchGCFunction && curHeapSize > heapSoftmax ) {
+  if (torchGCFunction && curHeapSize > heapSoftmax) {
     torchGCFunction(torchGCData);
-    long newHeapSize = THAtomicGetLong(&heapSize);
-    if(newHeapSize > heapSoftmax * heapSoftmaxGrowthThresh) {
+
+    // ensure heapSize is accurate before updating heapSoftmax
+    long newHeapSize = THAtomicAddLong(&heapSize, heapDelta) + heapDelta;
+    heapDelta = 0;
```

Comment:
nit: Maybe factor this out into its own function, something like `applyHeapDelta` or `updateHeapSize`? I know it's only two lines of code but it's easy to get wrong, e.g. forgetting to reset heapDelta.